README.md
![cx_health_sign](https://socialify.git.ci/panghaibin/cx_health_sign/image?description=1&forks=1&issues=1&language=1&stargazers=1)

**_注意：本项目对打卡的内容并不关心且一无所知，程序的原理仅仅是获取上一次的表单数据，再将其自动填写。程序不对使用本程序的用户负责，请自行保证表单的正确性与完整性。使用时请自觉遵守相应的约束性条款。_**

项目地址：<https://github.com/panghaibin/cx_health_sign>

最近更新日志：

[2.4.3] - 2022-09-21

适配 山东理工职业学院 健康打卡

适配 咸阳职业技术学院 早午晚检

修复 GitHub Actions 单用户消息推送不生效的 bug

......

[查看全部更新日志](CHANGELOG.md)

## 特色
- 支持在 GitHub Actions 上使用，以及自建服务器、PC等上使用

- 支持为多用户填报，单用户可填报多个表单，适配学校多个表单的情况

- 可支持学校各类自定义打卡表单

- 支持自助适配学校的表单模板（详见 [学习通健康打卡表单自适配指北](https://hbte.ch/1968.html) ）

- 支持[多种消息推送渠道](#消息推送介绍)，可将打卡上报结果推送给自己

- 支持将所有打卡结果推送给指定的一人

## 支持的学校表单
| 表单代码 | 名称 | 备注 |
| :---: | :---: | :---: |
| test  | 测试用表单，不限填报时间和次数 | [表单主页](http://office.chaoxing.com/front/web/apps/forms/fore/apply?id=13243&enc=3a9416c86432c5f667f2b23a88a0123a)
| default | 默认健康打卡表单 | [表单主页](http://office.chaoxing.com/front/web/apps/forms/fore/apply?id=7185&enc=f837c93e0de9d9ad82db707b2c27241e)
| nnnu | 南宁师范大学 两检 | [早检表单](http://office.chaoxing.com/front/web/apps/forms/fore/apply?id=99778&enc=5affca1a747445b8d3ec9de92612ecae) [午检表单](http://office.chaoxing.com/front/web/apps/forms/fore/apply?id=99783&enc=cb9894ce56b7e222cb3eab72d0fed834)
| hnucc | 湖南城建职业技术学院 学生健康信息填报 | [表单主页](https://office.chaoxing.com/front/third/apps/forms/fore/apply?id=86243&enc=de7939f413267efd9a0fd882dca9140b) |
| swut | 山东外国语职业技术大学 健康打卡 | [表单主页](https://office.chaoxing.com/front/web/apps/forms/fore/apply?id=139669&enc=d3fd2b1818f116a76aff41eee80ea348) |
| swut_2 | 山东外国语职业技术大学 午检打卡 | [表单主页](https://office.chaoxing.com/front/web/apps/forms/fore/apply?id=175235&enc=fb50b811a71a357bbb3a87424f7c074c) |
| hnisc | 湖南信息学院 健康打卡 | [表单主页](https://office.chaoxing.com/front/web/apps/forms/fore/apply?id=158324&enc=b08ae0de35d833ebc04ad7c5604f1b43) |
| xnec | 咸宁职业技术学院 学生晚检 | [表单主页](https://office.chaoxing.com/front/web/apps/forms/fore/apply?id=100992&enc=bd1883314d3b5f4b36c91dc1907b5c74) |
| xnec_2 | 咸宁职业技术学院 健康打卡-2 | [表单主页](https://office.chaoxing.com/front/web/apps/forms/fore/apply?id=100244&enc=c30d59556090a358fc0fb4992dd65cc1) |
| xnec_3 | 咸宁职业技术学院 学生早检 | [表单主页](https://office.chaoxing.com/front/web/apps/forms/fore/apply?id=100991&enc=67526ccf3ba6333c15c2a99a5b21682e) |
| xnec_4 | 咸宁职业技术学院 学生午检 | [表单主页](https://office.chaoxing.com/front/web/apps/forms/fore/apply?id=100989&enc=d854f35d6eb6ee0a700243ac1a4db86b) |
| qcuwh | 武汉晴川学院 健康打卡 | [表单主页](http://office.chaoxing.com/front/web/apps/forms/fore/apply?id=7185&enc=f837c93e0de9d9ad82db707b2c27241e) |
| hebart | 河北艺术职业学院 健康打卡 | [表单主页](http://office.chaoxing.com/front/web/apps/forms/fore/apply?id=886&enc=c8f7d4f5599544933f7c222b6b44e5c4) |
| cwxu | 南京信息工程大学滨江学院（无锡学院） 健康打卡 | [表单主页](http://office.chaoxing.com/front/web/apps/forms/fore/apply?id=3449&enc=627c625902a1fd27de56172186a3f903) |
| qvtu | 泉州职业技术大学 健康打卡 | [表单主页](http://office.chaoxing.com/front/web/apps/forms/fore/apply?id=200487&enc=b69dd988598ff61b4366fdb1f1962114) |
| hebau | 河北农业大学 健康上报 | [表单主页](http://office.chaoxing.com/front/web/apps/forms/fore/apply?id=204160&enc=a9b79f8b76307cd50a458b843d219ff2) |
| tust | 天津科技大学 本科健康打卡 | [表单主页](http://office.chaoxing.com/front/web/apps/forms/fore/apply?id=14673&enc=ed9e03b2050df7f56003dc0c0fa226d2) |
| wut | 武昌理工学院 本科健康打卡 | [表单主页](http://office.chaoxing.com/front/web/apps/forms/fore/apply?id=89398&enc=75244f75384287c902e57b080c4d1c6d) |
| whsw | 武汉生物工程学院 健康打卡 | [表单主页](http://office.chaoxing.com/front/web/apps/forms/fore/apply?id=174267&enc=2ea539d84f23c7852a021ed77008df9f) |
| sdpu | 山东理工职业学院 健康打卡 | [表单主页](http://office.chaoxing.com/front/web/apps/forms/fore/apply?id=211393&enc=4e585485d0b21ac1b15b6701b7c01f49) |

如果你的学校不在本项目支持列表之内，你可以：
- 自己抓包学习通的表单链接得到 `form id` 和 `enc` ，然后即可自助适配，详细步骤见 [学习通健康打卡表单自适配指北](https://hbte.ch/1968.html) 。适配好并测试通过后，即可向本项目 `Pull Request` 。
- 或者，直接提 Issue 请求适配

本项目支持使用 GitHub Actions 或在自建服务器上使用（阿里云等部分国内云服务IP可能被超星屏蔽，无法使用），通过使用 `crontab` 来定时开启。考虑到不同学校的打卡时间不一样，若使用 GitHub Actions 运行的，建议修改 `.github/workflows/report.yml` 内的定时时间（时区为**UTC时区**）；使用自建服务器的也一样，若为海外服务器也请注意服务器所使用的时区。

## 开始使用
### 方法一  在 GitHub Actions 上使用

1. `Star`并`Fork`本项目

2. 前往 `Settings`-`Secrets`-`Actions` ， 点击 `New repository secret`

3. 在新建 Secret 的界面上， `Name` 值输入 `NEWUSERS` ， `Value` 处输入用户的配置信息，多个用户配置之间使用分号隔开。

    > 在旧版本（1.4.0及以前）中采用的`Name`值为`USERS`，格式也难以阅读。现在的版本`Name`为`NEWUSERS`，并采用如下的便于阅读的设置方案，旧的方案仍然兼容。

    单个用户的配置信息分别由几对`key=value`组成，格式如下（[ ]内为可选参数）：
    ```
    un=username,pd=password,pt=post_type[,si=school_id,at=api_type,ak=api_key]
    ```

    `username`, `password` 分别是用户名（学号或手机号）和密码

    `post_type`是本程序目前支持填报的表单代码，详情见[支持的学校表单](#支持的学校表单)中的[表单代码]一栏。如果需要同时填报多个表单，请使用`|`符号连接不同的表单。如 `swut|swut_2`

    `school_id`是学校代号（仅在学号登录时需要），获取方法见[学校id获取](#学校id获取)

    `api_type`, `api_key`是消息推送类型代号及密钥（可选），详情见[消息推送介绍](#消息推送介绍)

    示例如下：

    使用手机号登录，表单代码为`nnnu`，不需要消息推送，仅需要基础填报功能：

    ```
    un=138xxxx,pw=123xxx,pt=nnnu
    ```

    使用手机号登录，需要同时使用表单代码`swut`和`swut_2`，需要消息推送服务：
    ```
    un=139xxxx,pw=567yyy,pt=swut|swut_2,at=2,ak=xxxx
    ```

    使用学号登录（查询得到学校代码为`209`），表单代码为`default`，需要消息推送服务：

    ```
    un=2019xxxx,pw=789zzz,pt=default,si=209,at=2,ak=xxxx
    ```

    以上几个`key=value`对的顺序可以交换，要注意末尾不能是`,`

    为多个用户设置时，每个用户的配置信息之间，用分号间隔。例如上面三个例子合在一起时：

    ```
    un=138xxxx,pw=123xxx,pt=nnnu;un=139xxxx,pw=567yyy,pt=swut|swut_2,at=2,ak=xxxx;un=2019xxxx,pw=789zzz,pt=default,si=209,at=2,ak=xxxx
    ```

    同样注意末尾不需要有分号`;`

    完成后点击`Add secret`保存该 Secret

4. （可选）用同样的方法添加一个名为 `NEWSEND` 的 Secret ，用于将所有用户的填报结果推送给管理员。

    > 在旧版本（1.4.0及以前）中采用的`Name`值为`SEND`，格式也难以阅读。现在的版本`Name`为`NEWSEND`，并采用如下的便于阅读的设置方案，旧的方案仍然兼容。

    格式为`at=api_type,ak=api_key`，例如

    ```
    at=2,ak=xxxx
    ```

5. （可选）添加一个名为 `LOGPASS` 的 Secret ，用于将 GitHub Actions 的执行结果日志加密

    设置该 Secret 后会将日志文件用 7z 加密压缩并上传，可在每个 Action 页面的 Artifacts 处下载，使用任意解压软件解压即可查看。

    若不设置该 Secret 则日志直接输出在 Actions 的执行结果页

6. （可选）添加一个名为 `SLEEPTIME` 的 Secret ，以在填报时控制休眠时间

    设置值为 `random` ，每次会随机休眠 30 ~ 360 秒，若不设置默认休眠 5 秒。

7. Secret 添加完成后，前往项目的 `Actions` 面板，同意开启并进入 Actions 。然后选择 `Health Report` ，点击 `Enable workflow` 开启工作流。此时 Actions 开启成功，可以点击 `Run workflow` 测试填报一次。

8. 当本项目更新时，你所 Fork 的项目不会自动更新。可以选择安装 GitHub Apps 中的 [Pull App](https://github.com/apps/pull) ，保持你的 Fork 始终最新。安装时默认会应用到所有项目，建议手动选择需要应用的项目。注意该 App 会强制覆盖你对 Fork 项目的操作，如果你有改动（例如修改了 `.github/workflows/report.yml` ），请注意备份。

    若不安装，需要手动更新时，在你的项目主页上点击 `Fetch upstream`-`Fetch and merge` 即可更新程序。

### 方法二  在自己的服务器上使用

1. 克隆本项目，进入项目文件夹
    ```shell
    git clone https://github.com/panghaibin/cx_health_sign
    cd cx_health_sign
    # 安装依赖
    pip3 install -r requirements.txt
    ```

2. 添加用户
    ```shell
    python3 main.py add
    ```

    部分参数选填，根据提示执行添加用户操作

    获取学校id的方法见[学校id获取](#学校id获取)

    要使用消息推送服务，你需要提前注册第三方的消息推送服务以获取密钥，见[消息推送介绍](#消息推送介绍)

3. （可选）设置全局推送

    ```shell
    python3 main.py send
    ```

    根据提示操作，设置后所有用户的填报结果都会推送到对应接口


以上两步也可通过直接创建 `setting.yaml` 文件以保存配置，格式参考 `setting.bak.yaml` 。

4. 执行程序
    ```shell
    python3 main.py
    ```

    这将会立即填报一次，确保可以正确填报


5. 设置定时填报
    ```shell
    vim /etc/crontab
    ```

    向其中添加
    ```
    0 7,12,19 * * * root /usr/bin/python3 /root/cx_health_sign/main.py >> /root/cx_health_sign/output.log
    ```
    此仅为示意，请根据实际情况作出修改，例如将时间设置正确，脚本的路径正确

    程序在填报时每次填报默认休眠 5s，如果需要自定义休眠时间，如每次休眠 30s：

    ```
    0 7,12,19 * * * root export sleep_time=30 && /usr/bin/python3 /root/cx_health_sign/main.py >> /root/cx_health_sign/output.log
    ```
   
    另外还支持随机休眠30~360s，设置为 `sleep_time=random` 即可

    保存文件后执行
    ```shell
    crontab /etc/crontab
    ```
    以应用 crontab 配置

## 消息推送介绍
目前支持以下消息推送服务：

| 接口代号 | 名称| 官网 |
| :---: | :---: | :---: |
| 1 | Server酱 | https://sct.ftqq.com/ |
| 2 | 推送加 | https://www.pushplus.plus/ |
| 3 | ~~推送加（hxtrip域名下）~~ | ~~https://pushplus.hxtrip.com/~~ 该服务已下线 |

请前往任意官网注册得到`key`后即可在本项目中使用，在 GitHub Actions 中使用时注意接口代号正确设置。

## 学校id获取
1. 访问 http://passport2.chaoxing.com/login

2. 按下面动图操作

    ![2020/04/15/cdf5a0415014614.gif](http://cdn.z2blog.com/2020/04/15/cdf5a0415014614.gif)

    [图片来源](https://github.com/mkdir700/chaoxing_auto_sign/blob/latest/api/readme.md)

## Thanks
本项目受到了[mkdir700/chaoxing_auto_sign](https://github.com/mkdir700/chaoxing_auto_sign) 的启发，特此表示感谢！


CHANGELOG.md
# 更新日志

## [2.4.3] - 2022-09-21
### Added
- 适配 山东理工职业学院 健康打卡

## [2.4.2] - 2022-09-03
### Added
- 适配 咸阳职业技术学院 早午晚检

## [2.4.1] - 2022-09-01
### Fixed
- 修复 GitHub Actions 单用户消息推送不生效的 bug

## [2.3.9] - 2022-08-10
### Added
- 适配 武汉生物工程学院 健康打卡

## [2.3.8] - 2022-07-30
### Added
- 适配 武昌理工学院 健康打卡

## [2.3.7] - 2022-07-20
### Added
- 适配 咸宁职业技术学院 健康打卡-2

## [2.3.6] - 2022-06-22
### Added
- 适配 天津科技大学 本科健康打卡

## [2.3.5] - 2022-06-09
### Added
- 适配 河北农业大学 健康上报

## [2.3.4] - 2022-05-31
### Removed
- 因官方停用，移除 `推送加（hxtrip域名下）` 推送接口的支持

## [2.3.3] - 2022-05-27
### Added
- 适配 泉州职业技术大学 健康打卡

## [2.3.2] - 2022-04-11
### Fixed
- 修复休眠时间bug

## [2.3.1] - 2022-04-11
### Fixed
- 修复 表单存在多个体温值的情况

## [2.3.0] - 2022-04-11
### Added
- 适配 南京信息工程大学滨江学院（无锡学院） 健康打卡

## [2.2.0] - 2022-04-10
### Added
- 适配 河北艺术职业学院 健康打卡

## [2.1.0] - 2022-04-05
### Added
- 新增 自助适配健康填报表单功能，详见 [https://hbte.ch/1968.html](https://hbte.ch/1968.html)

## [2.0.2] - 2022-04-04
### Fixed
- 修复 咸宁职业技术学院 健康打卡
- 优化了部分代码

## [2.0.1] - 2022-04-04
### Added
- 优化休眠机制输出

## [2.0.0] - 2022-04-04
### Added
- 现支持单用户同时填报多个表单，详见 README.md
### Docs
- 增加休眠机制的文档说明

## [1.4.1] - 2022-04-03
### Changed
- 改善 GitHub Secrets 的配置方案
- 修改 GitHub Actions 日志输出方案，可直接查看
- 填报结果用户名将部分打码

## [1.4.0] - 2022-03-17
### Added
- 适配 山东外国语职业技术大学 午检打卡
- 适配 咸宁职业技术学院 健康打卡
- 适配 武汉晴川学院 健康打卡

## [1.3.2] - 2022-03-07
### Fixed
- 修复 南宁师范大学

## [1.3.1] - 2022-02-08
### Added
- 增加 调试堆栈输出
### Changed
- 修改 休眠函数逻辑 默认不休眠
### Fixed
- 修复 山东外国语职业技术大学 健康打卡

## [1.3.0] - 2022-02-01
### Added
- 适配 湖南信息学院 健康打卡

## [1.2.0] - 2022-01-04
### Added
- 适配 山东外国语职业技术大学 健康打卡

## [1.1.0] - 2021-12-10
### Added
- 适配 湖南城建职业技术学院 学生健康信息填报
- 增加 填报休眠机制
### Fixed
- 修复 消息发送函数bug

## [1.0.2] - 2021-10-03
### Added
- 新增 GitHub Actions 输出加密
### Fixed
- 修复 GitHub Actions 执行时间

## [1.0.1] - 2021-10-02
### Added
- 新增 命令行设置全局推送
- 新增 GitHub Actions 使用

## [1.0.0] - 2021-09-28
### Added
- 基础功能实现
- 适配 南宁师范大学 三检打卡


main.py
# -*- coding: utf8 -*-
import os
import sys
import json
import random
import traceback
import requests
import logging
from time import sleep
from setting import Setting
from setting import GitHub
from config import Time
from config.default import DefaultHealthReport
from config.nnnu import NNNUHealthReport
from config.hnucc import HNUCCHealthReport
from config.swut import SWUTHealthReport
from config.swut_2 import SWUTHealthReportNoon
from config.hnisc import HNISCHealthReport
from config.xnec import XNECHealthReport
from config.xnec_2 import XNECHealthReport2
from config.xnec_3 import XNECHealthReport3
from config.xnec_4 import XNECHealthReport4
from config.qcuwh import QCUWHHealthReport
from config.hebart import HEBARTHealthReport
from config.cwxu import CWXUHealthReport
from config.qvtu import QVTUHealthReport
from config.hebau import HEBAUHealthReport
from config.test import TestReport
from config.tust import TUSTHealthReport
from config.wut import WUTHealthReport
from config.whsw import WHSWHealthReport
from config.sdpu import SDPUHealthReport


class MainHandle(object):
    """
    主函数
    """

    def __init__(self):
        self._users: list = []
        self.users_num: int = 0
        self._global_api: dict = {}
        self._reporters = {
            'test': TestReport,
            'default': DefaultHealthReport,
            'nnnu': NNNUHealthReport,
            'hnucc': HNUCCHealthReport,
            'swut': SWUTHealthReport,
            'swut_2': SWUTHealthReportNoon,
            'hnisc': HNISCHealthReport,
            'xnec': XNECHealthReport,
            'xnec_2': XNECHealthReport2,
            'xnec_3': XNECHealthReport3,
            'xnec_4': XNECHealthReport4,
            'qcuwh': QCUWHHealthReport,
            'hebart': HEBARTHealthReport,
            'cwxu': CWXUHealthReport,
            'qvtu': QVTUHealthReport,
            'hebau': HEBAUHealthReport,
            'tust': TUSTHealthReport,
            'wut': WUTHealthReport,
            'whsw': WHSWHealthReport,
            'sdpu': SDPUHealthReport,
        }

        # 健康上报结果，多用户存储在一个数组
        self.report_results: list = []
        # 消息发送结果
        self.send_results: list = []
        # 全局消息发送结果
        self.global_send_result: str = ''

    @staticmethod
    def _sleep():
        """
        休眠函数，多个用户填报时，防止被封
        通过设置环境变量控制，默认 5s
        """
        os_sleep = os.getenv('sleep_time', '')
        os_sleep = 5 if os_sleep == '' else os_sleep
        if os_sleep == 'random':
            sleep_time = random.randint(30, 360)
        else:
            sleep_time = int(os_sleep)
        print('休眠 {}s'.format(sleep_time))
        for i in range(sleep_time, 0, -10):
            print("休眠剩余%s秒" % i)
            sleep(10 if i > 10 else i)
        print("休眠结束")

    def main(self):
        setting = Setting()
        self._users = setting.get_users(post_type=None)
        self.users_num = len(self._users)
        self._global_api = setting.global_api

        if self.users_num == 0:
            print('用户数量为0，开始添加用户')
            self.add_user()
            print('可执行 `python main.py add` 或修改 setting.yaml 文件继续添加')
            print('开始填报测试')
        # 填报全部用户
        self.report_all()
        # 发送全局消息
        self.global_send()

        # 打印每个用户健康报告填报结果
        print(self.report_results)
        # 打印每个用户的消息发送结果
        print(self.send_results)
        # 打印全局消息发送结果
        print(self.global_send_result)

    def report_all(self):
        for user in self._users:
            post_types = user['post_type']
            for post_type in post_types:
                self._sleep()
                try:
                    r = self._reporters[post_type](user['username'], user['password'], user.get('school_id', ''))
                    t = r.report()
                    self.report_results.append(t)
                except Exception as e:
                    logging.exception(e)
                    msg = traceback.format_exc()
                    self.report_results.append(str(msg))
                if user.get('api_type', 0) != 0:
                    send = SendMsg(user, result=self.report_results[-1])
                    self.send_results.append(send.send_result)
                else:
                    self.send_results.append('用户未指定消息推送服务')

    def global_send(self) -> str:
        api = {
            'api_type': self._global_api.get('api_type', 0),
            'api_key': self._global_api.get('api_key', '')
        }
        if api['api_type'] != 0:
            send = SendMsg(api, result_list=self.report_results)
            self.global_send_result = send.send_result
        else:
            self.global_send_result = '未指定全局消息推送'
        return self.global_send_result

    @staticmethod
    def _input(message: str, is_require=True, message_list=None):
        """
        封装的输入函数，判断是否必填项，支持对一个列表的选择（返回该列表的下标 (int) ），进行了错误处理
        """
        if message_list is not None:
            for i in range(len(message_list)):
                print("%s: %s" % (i + 1, message_list[i]))

        while True:
            data = input(message)
            if data == '' and not is_require:
                break
            elif data != '' and message_list is None:
                break
            elif data != '' and message_list is not None:
                try:
                    data = int(data) - 1
                    if 0 <= data <= len(message_list) - 1:
                        print('已选择: %s' % message_list[data])
                        break
                except Exception as e:
                    print(e)
        return data

    def add_user(self) -> bool:
        setting = Setting()

        print('输入用户名')
        print('手机号/邮箱/学号均可，必填')
        username = self._input('请输入: ')

        print('输入密码，必填')
        password = self._input('请输入: ')

        print('输入需要打卡的配置类型，必填')
        report_types = list(self._reporters.keys())
        post_type = []
        while True:
            input_value = self._input('请选择: ', message_list=report_types)
            input_value = report_types[input_value]
            post_type.append(input_value)
            print('是否继续添加更多表单？')
            to_exit = self._input('请选择：', message_list=['是', '否'])
            if to_exit == 1:
                break
        post_type = list(set(post_type))

        print('输入学校id，选填 若使用学号登录则必填')
        school_id = self._input('请输入: ', is_require=False)

        print('选择消息推送服务，选填')
        send = SendMsg
        send_types = [send.api_types_name[i] + ' ' + send.api_types_url[i] for i in range(len(send.api_types_name))]
        api_type = self._input('请选择: ', is_require=False, message_list=send_types[1:])
        while api_type in [2]:
            print('***该接口已弃用，请选择其它接口！')
            api_type = self._input('请选择: ', is_require=False, message_list=send_types[1:])
        api_type = 0 if api_type == '' else api_type + 1

        api_key = ''
        if api_type != 0:
            print('输入消息推送 key')
            api_key = self._input('请输入: ')

        add_result = setting.add_user(username, password, post_type, school_id, api_type, api_key)
        if add_result:
            print('用户添加成功')
            self._users = setting.get_users(post_type=None)
        else:
            print('用户添加失败')
        return add_result

    def set_global_send(self) -> bool:
        setting = Setting()
        print('设置全局推送服务')

        print('选择消息推送服务')
        send = SendMsg
        send_types = [send.api_types_name[i] + ' ' + send.api_types_url[i] for i in range(len(send.api_types_name))]
        api_type = self._input('请选择: ', is_require=True, message_list=send_types[1:])
        while api_type in [2]:
            print('***该接口已弃用，请选择其它接口！')
            api_type = self._input('请选择: ', is_require=False, message_list=send_types[1:])
        api_type = 0 if api_type == '' else api_type + 1

        print('输入消息推送 key')
        api_key = self._input('请输入: ')
        add_result = setting.set_global_send(api_type, api_key)
        if add_result:
            print('全局推送设置成功')
            self._global_api = setting.global_api
        else:
            print('全局推送设置失败')
        return add_result


class GitHubHandle(MainHandle):

    def __init__(self):
        super().__init__()

    def main(self):
        setting = GitHub()
        self._users = setting.get_users(post_type=None)
        self.users_num = len(self._users)
        self._global_api = setting.global_api

        if self.users_num == 0:
            raise Exception('GitHub Actions Secrets 未设置')
        # 填报全部用户
        self.report_all()
        # 发送全局消息
        self.global_send()

        # 打印每个用户健康报告填报结果
        print(self.report_results)
        # 打印每个用户的消息发送结果
        print(self.send_results)
        # 打印全局消息发送结果
        print(self.global_send_result)

    def add_user(self) -> bool:
        return False

    def set_global_send(self) -> bool:
        return False


class SendMsg(object):
    """
    消息发送类
    """

    api_types_name = [
        '未设置',
        'Server酱',
        '推送加',
        '(X)[该接口已停用] 推送加(hxtrip域下)'
    ]

    api_types_url = [
        '',
        'https://sct.ftqq.com/',
        'https://www.pushplus.plus/',
        'https://pushplus.hxtrip.com/'
    ]

    def __init__(self, user_api: dict, result: str = None, result_list: list = None):
        """
        result: 仅发送单条消息
        result_list: 发送多条消息，合并在消息的详情发送
        两个二选一， result 优先
        """
        if not result and not result_list:
            raise Exception('必须传入 result 或 result_list')

        self.api_type = user_api.get('api_type', 0)
        self.api_key = user_api.get('api_key', '')
        if self.api_type not in range(1, len(self.api_types_name)):
            raise Exception('未正确配置消息发送类型')
        self.api_type_name = self.api_types_name[self.api_type]
        self.api_type_url = self.api_types_url[self.api_type]
        self.send_result: dict = {}
        self.send_result_bool: bool = False

        self.result_list = result_list
        self.title = result_list[0] if not result else result
        self.title = self.title.replace("\n", '<br>')[:99]
        self.desp: str = Time().report_time_sec + "\n\n"
        self.desp += "\n\n".join(result_list) if not result else result
        # 添加随机字符以确保提交成功
        self.desp += "\n\n" + str(random.randint(0, 100000))

        # 这里要写在最后，确保 title resp 已初始化
        self._send_api = (None, self.server_chan, self.push_plus, self.push_plus_hxtrip)

        self.send_msg()

    def server_chan(self) -> bool:
        url = "http://sctapi.ftqq.com/%s.send" % self.api_key
        data = {
            'text': self.title,
            'desp': self.desp
        }
        text = requests.post(url, data=data).text
        send_result = json.loads(text)
        self.send_result = send_result
        if send_result['code'] == 0:
            return True
        else:
            return False

    def push_plus(self) -> bool:
        url = "http://www.pushplus.plus/send"
        data = {
            'token': self.api_key,
            'title': self.title,
            'content': self.desp,
            'template': 'markdown'
        }
        headers = {'Content-Type': 'application/json'}
        body = json.dumps(data).encode(encoding='utf-8')
        text = requests.post(url, data=body, headers=headers).text
        send_result = json.loads(text)
        self.send_result = send_result
        if send_result['code'] == 200:
            return True
        else:
            return False

    @staticmethod
    def push_plus_hxtrip() -> bool:
        print('该消息推送接口已弃用，请更换其它接口')
        return False

    def send_msg(self) -> bool:
        result_bool = self._send_api[self.api_type]()
        self.send_result_bool = result_bool
        return self.send_result_bool


if __name__ == '__main__':
    if len(sys.argv) == 1:
        MainHandle().main()
    elif len(sys.argv) == 2:
        if sys.argv[1] == 'add':
            MainHandle().add_user()
        elif sys.argv[1] == 'send':
            MainHandle().set_global_send()
        elif sys.argv[1] == 'gh':
            GitHubHandle().main()
